app/services/DataPointMapper.java

```diff
@@ -47,18 +46,27 @@ public void mapDataPoints(List<String> jsonDataPoints) throws DatapointMappingEx
         tx.begin();
         try (DatapointParser parser = datapointParserSupplier.get()) {
 
-            Set<UUID> datasetIds = new HashSet<>();
+            Map<UUID, Integer> datasetCounts = new HashMap<>();
             for (String record : jsonDataPoints) {
                 logger.debug("Processing data point: {}", record);
 
                 final DataPointRecord dataPointRecord = parseDataPointRecord(record);
-                datasetIds.add(dataPointRecord.getDatasetID());
                 mapDataPoint(parser, dataPointRecord, entityManager);
+                UUID datasetID = dataPointRecord.getDatasetID();
+                datasetCounts.put(datasetID, datasetCounts.computeIfAbsent(datasetID, k-> 0) + 1);
```

Comment:
You can use `getOrDefault` rather than `computeIfAbsent` if the default is a constant.